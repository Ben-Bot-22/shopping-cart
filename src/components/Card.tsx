import { Link } from 'react-router-dom'

type Props = {
  source: string
  name: string
  price: number
}

function Card({ source, name, price }: Props) {
  return (
    <Link to="/product" state={{ source, name, price }}>
      <div className="bold flex h-80 w-52 flex-col items-center rounded-lg border-4 border-accent bg-accent text-white shadow-lg shadow-zinc-500 hover:border-white">
        <img
          src={source}
          alt={name}
          className="h-64 w-52 rounded-t object-fill"
        ></img>
        <h1>{name}</h1>
        <h2>${price}</h2>
      </div>
    </Link>
  )
}

export default Card
