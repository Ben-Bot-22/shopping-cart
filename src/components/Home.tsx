import react from 'react'
import { Link } from 'react-router-dom'

function Home() {
  return (
    <>
      <div className="flex flex-col items-center gap-6">
        <img
          src="/src/assets/catHero.jpg"
          alt="cat walking next to bay"
          className="w-px-600 h-auto border-4 border-first"
        ></img>
        <Link
          to="/shop"
          className="absolute top-[50%] flex h-11 w-40 items-center justify-center rounded-3xl border-4 bg-accent text-xl text-white"
          data-testid="shop-button"
        >
          <div>Shop Now</div>
        </Link>
      </div>
    </>
  )
}

export default Home
