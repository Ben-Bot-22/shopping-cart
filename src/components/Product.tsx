import react from 'react'
import { useLocation, Link } from 'react-router-dom'

/*
- item page
-- proceed to checkout (if something is in cart)
*/

function Product() {
  const location = useLocation()
  return (
    <div className="mt-4 flex justify-center">
      <div className="border-accent flex h-full items-center justify-between rounded-xl border-4 bg-white">
        <img
          src={location.state.source}
          alt={location.state.text}
          className="border-accent h-5/6 w-80 rounded-l-lg border-r-4 object-fill"
        />
        <div className="border-accent flex flex-col items-center gap-4 p-8">
          <h1 className="text-accent text-2xl font-bold">
            {location.state.text}
          </h1>
          <h1 className="text-accent text-2xl">{location.state.price}</h1>
          <Link
            className="bg-accent text-bold flex h-8 w-44 items-center justify-center rounded-md text-xl text-white"
            to="/cart"
          >
            <span>Add to Cart</span>
          </Link>
          <Link
            className="text-bold text-accent border-accent flex h-8 w-44 items-center justify-center rounded-md border-2 text-xl"
            to="/cart"
          >
            <span>View in Cart</span>
          </Link>
        </div>
      </div>
    </div>
  )
}

export default Product
