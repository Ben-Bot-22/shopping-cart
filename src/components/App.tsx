import { useState } from 'react'
import { Routes, Route } from 'react-router-dom'
import Nav from './Nav'
import Home from './Home'
import Shop from './Shop'
import Cart from './Cart'
import Product from './Product'
import { v4 as uuidv4 } from 'uuid'
import { cartItemData } from '../utils/types'

// // function addToCart(
// //   id: string,
// //   name: string,
// //   price: number,
// //   quantity: number,
// //   src: string
// // ): void {
// //   // find item
// //   const item = cartItems.find((item) => item.id === id)
// //   // if entry exists, update quantity
// //   if (item) {
// //     // update quantity
// //     updateQuantity(id, item.quantity + quantity)
// //   } else {
// //     // else, add entry
// //     addNewItem(name, price, quantity, src)
// //   }
// // }

function App() {
  const [cartItems, setCartItems] = useState(cartItemData)

  const updateQuantity = (id: string, quantity: number) => {
    const newCartItems = cartItems.map((item) => {
      if (item.id === id) {
        return { ...item, quantity: quantity }
      }
      return item
    })
    setCartItems(newCartItems)
  }

  const addNewItem = (
    name: string,
    price: number,
    quantity: number,
    src: string
  ) => {
    const newCartItems = [
      ...cartItems,
      { id: uuidv4(), name, price, quantity, imageSource: src }
    ]
    setCartItems(newCartItems)
  }

  const removeFromCart = (id: string) => {
    // remove from cart
    setCartItems(cartItems.filter((item) => item.id !== id))
  }

  return (
    <div>
      <Nav />
      <Routes>
        <Route path="/" element={<Home />} />
        <Route path="/shop" element={<Shop />} />
        <Route
          path="/cart"
          element={
            <Cart
              cartItems={cartItems}
              updateQuantity={updateQuantity}
              addNewItem={addNewItem}
              removeFromCart={removeFromCart}
            />
          }
        />
        <Route
          path="/product"
          element={
            <Product
              cartItems={cartItems}
              updateQuantity={updateQuantity}
              addNewItem={addNewItem}
              removeFromCart={removeFromCart}
            />
          }
        />
        <Route path="*" element={<h1>404 Not Found</h1>} />
      </Routes>
    </div>
  )
}

export default App

/*
git hub commit
Fix mismatch tailwind ordering in config

INTERACTIVE

- shopping cart: number of items in cart
-- up / down
-- input field: manually type

- navigation bar: number of items in cart

- cart
-- update quantity
-- manually type quantity in input field
-- pay now
-- update total price

- product page: view in cart (only if product is in cart)

TESTING
Unit test
- item: quantity up / down
- item: remove

- Setup tests with React Testing Library
-- snapshot
-- screen
*/
