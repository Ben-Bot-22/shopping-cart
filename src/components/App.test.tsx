import React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom' // optional
import userEvent from '@testing-library/user-event'
import App from './App'
import Product from './Product'
import Home from './Home'
import Card from './Card'
import { BrowserRouter, Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'

/*
TODO:

** TESTS **

Add to cart button works

SHOP
- add to cart adds view in cart button
- adding cats to cart; cart has cats

NAVBAR
- add to cart updates nav bar quantity
- remove from cart update nav bar quantity

CART
- remove button removes item
- quantity updates total price
- Up / down buttons update quantity

*/

const MockApp = () => {
  return (
    <BrowserRouter>
      <Home />
    </BrowserRouter>
  )
}

describe('Add to cart button populates cart screen', () => {
  it('renders shop now button', async () => {
    render(<MockApp />)
    const shopNowButton = screen.getByRole('link', { name: /Shop Now/i })
    expect(shopNowButton).toBeInTheDocument()
  })

  it('renders correctly', () => {
    const snap = render(<MockApp />)
    expect(snap).toMatchSnapshot()
  })
  // it('renders view in cart button', async () => {
  //   render(<MockApp />)
  //   const shopNowButton = screen.getByRole('link', { name: /Shop Now/i })
  //   userEvent.click(shopNowButton)
  //   screen.debug()
  //   const addToCartButton = screen.getByRole('button', { name: /Add to Cart/i })
  //   userEvent.click(addToCartButton)
  //   const viewCartButton = screen.getByRole('button', { name: /View in Cart/i })
  //   expect(viewCartButton).toBeInTheDocument()
  // })
})
