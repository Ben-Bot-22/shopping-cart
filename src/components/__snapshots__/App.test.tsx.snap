// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add to cart button populates cart screen renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col items-center gap-6"
      >
        <img
          alt="cat walking next to bay"
          class="w-px-600 h-auto border-4 border-first"
          src="/src/assets/catHero.jpg"
        />
        <a
          class="absolute top-[50%] flex h-11 w-40 items-center justify-center rounded-3xl border-4 bg-accent text-xl text-white"
          data-testid="shop-button"
          href="/shop"
        >
          <div>
            Shop Now
          </div>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col items-center gap-6"
    >
      <img
        alt="cat walking next to bay"
        class="w-px-600 h-auto border-4 border-first"
        src="/src/assets/catHero.jpg"
      />
      <a
        class="absolute top-[50%] flex h-11 w-40 items-center justify-center rounded-3xl border-4 bg-accent text-xl text-white"
        data-testid="shop-button"
        href="/shop"
      >
        <div>
          Shop Now
        </div>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
